File: reconer/src/motion-detect.ts
/**
 * motion-detect.ts — Step 1: 从速拧视频切分 静止/转动 片段 (对应 motion_detect.py)。
 *
 * 两层:
 *   - segmentFromDiffs(diffs, fps): 纯状态机 (阈值+回滞→聚合→过滤短段→合并), 可单测,
 *     并修掉原 Python 空输入即 IndexError 的 bug。
 *   - computeFrameDiffs(video): ffmpeg 单 pass 流式算 ROI 灰度帧差均值喂给状态机。
 *
 * 注意: ffmpeg 的 gray/gblur 与 cv2 的 BGR2GRAY/GaussianBlur 近似不同, 帧差 *幅值* 与
 * cv2 略有出入 (同 B 面色的 H.264 解码差), 但状态机逻辑与 Python 逐值一致 (parity-motion.ts)。
 */
import { spawn, spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";

// ============================================================
// 配置 (与 motion_detect.py 完全一致)
// ============================================================

/** 魔方 ROI (归一化坐标, 相对整帧): 左上 (X1,Y1) 右下 (X2,Y2) */
export const ROI_X1 = 0.25, ROI_Y1 = 0.15;
export const ROI_X2 = 0.75, ROI_Y2 = 0.85;

export const MOTION_THRESHOLD = 5.0; // 高于此 → 运动帧
export const STATIC_THRESHOLD = 3.0; // 低于此 → 静止帧 (回滞防抖)
export const MIN_MOVE_FRAMES = 3; // 过短运动段视为噪声
export const MIN_STATIC_FRAMES = 5; // 过短静止间隔并入同一动作

export type SegType = "STATIC" | "MOVING";

export interface Segment {
  type: SegType;
  startFrame: number;
  endFrame: number;
  startTime?: number;
  endTime?: number;
}

// ============================================================
// 纯状态机 (可单测)
// ============================================================

/**
 * 由逐帧差值序列切分片段。diffs[i] = 帧 i 与 i+1 的 ROI 灰度绝对差均值
 * (即 diffs 从第 1 帧起, 与 motion_detect.py 同)。空输入返回 []。
 */
export function segmentFromDiffs(diffs: number[], fps: number): Segment[] {
  if (diffs.length === 0) return []; // 修 Python 的空输入 IndexError

  // 逐帧状态 (带回滞)
  const states: SegType[] = [];
  let state: SegType = "STATIC";
  for (const diff of diffs) {
    if (state === "STATIC") {
      if (diff > MOTION_THRESHOLD) state = "MOVING";
    } else {
      if (diff < STATIC_THRESHOLD) state = "STATIC";
    }
    states.push(state);
  }

  // 聚合为片段 (startFrame = startIdx+1, 因 diffs 从第 1 帧起)
  const rawSegments: Segment[] = [];
  let currState = states[0];
  let startIdx = 0;
  for (let i = 1; i < states.length; i++) {
    if (states[i] !== currState) {
      rawSegments.push({ type: currState, startFrame: startIdx + 1, endFrame: i });
      currState = states[i];
      startIdx = i;
    }
  }
  rawSegments.push({ type: currState, startFrame: startIdx + 1, endFrame: states.length });

  // 过短运动段 → 静止
  const filtered: Segment[] = rawSegments.map((seg) => {
    const duration = seg.endFrame - seg.startFrame;
    return seg.type === "MOVING" && duration < MIN_MOVE_FRAMES ? { ...seg, type: "STATIC" as const } : seg;
  });

  // 合并相邻同类型
  const merged: Segment[] = [{ ...filtered[0] }];
  for (const seg of filtered.slice(1)) {
    const last = merged[merged.length - 1];
    if (seg.type === last.type) last.endFrame = seg.endFrame;
    else merged.push({ ...seg });
  }

  // 过短静止间隔 → 运动 (需前面已有段)
  const filtered2: Segment[] = [];
  for (const seg of merged) {
    const duration = seg.endFrame - seg.startFrame;
    if (seg.type === "STATIC" && duration < MIN_STATIC_FRAMES && filtered2.length) seg.type = "MOVING";
    filtered2.push(seg);
  }

  // 再次合并
  const final: Segment[] = [{ ...filtered2[0] }];
  for (const seg of filtered2.slice(1)) {
    const last = final[final.length - 1];
    if (seg.type === last.type) last.endFrame = seg.endFrame;
    else final.push({ ...seg });
  }

  // 时间信息
  for (const seg of final) {
    seg.startTime = seg.startFrame / fps;
    seg.endTime = seg.endFrame / fps;
  }
  return final;
}

// ============================================================
// ffmpeg 帧差喂料
// ============================================================

interface VideoMeta {
  fps: number;
  width: number;
  height: number;
}

/** 用 ffprobe 取 帧率/宽/高。 */
export function probeVideo(videoPath: string, ffprobePath = "ffprobe"): VideoMeta {
  const args = [
    "-v", "error", "-select_streams", "v:0",
    "-show_entries", "stream=width,height,r_frame_rate",
    "-of", "json", videoPath,
  ];
  const res = spawnSync(ffprobePath, args, { encoding: "utf8" });
  if (res.status !== 0) throw new Error(`ffprobe failed (${res.status}): ${res.stderr || res.error}`);
  const stream = (JSON.parse(res.stdout).streams ?? [])[0];
  if (!stream) throw new Error(`ffprobe: ${videoPath} 无视频流`);
  const [num, den] = String(stream.r_frame_rate).split("/").map(Number);
  const fps = den ? num / den : num;
  if (!Number.isFinite(fps) || fps <= 0) throw new Error(`ffprobe: 非法帧率 "${stream.r_frame_rate}"`);
  return { fps, width: stream.width, height: stream.height };
}

/** ROI 归一化坐标 → 像素裁剪框 (int 截断, 与 Python 一致)。 */
export function roiCrop(width: number, height: number): { x: number; y: number; w: number; h: number } {
  const x1 = Math.floor(width * ROI_X1), y1 = Math.floor(height * ROI_Y1);
  const x2 = Math.floor(width * ROI_X2), y2 = Math.floor(height * ROI_Y2);
  return { x: x1, y: y1, w: x2 - x1, h: y2 - y1 };
}

/**
 * ffmpeg 单 pass 流式计算逐帧 ROI 灰度帧差均值 (仅驻留 2 帧, 支持长 4K 视频)。
 * 顺序: format=gray → gblur(sigma≈1.1, 对应 cv2 GaussianBlur 5×5) → crop 到 ROI。
 */
export function computeFrameDiffs(
  videoPath: string,
  ffmpegPath = "ffmpeg",
  ffprobePath = "ffprobe",
): Promise<{ diffs: number[]; fps: number; width: number; height: number }> {
  const meta = probeVideo(videoPath, ffprobePath);
  const crop = roiCrop(meta.width, meta.height);
  const frameBytes = crop.w * crop.h;
  const vf = `format=gray,gblur=sigma=1.1,crop=${crop.w}:${crop.h}:${crop.x}:${crop.y}`;
  const args = [
    "-hide_banner", "-v", "error", "-i", videoPath,
    "-vf", vf, "-fps_mode", "passthrough",
    "-pix_fmt", "gray", "-f", "rawvideo", "pipe:1",
  ];

  return new Promise((resolve, reject) => {
    const child = spawn(ffmpegPath, args);
    const diffs: number[] = [];
    let prev: Uint8Array | null = null;
    let pending: Buffer = Buffer.alloc(0);
    let stderr = "";

    child.stdout.on("data", (chunk: Buffer) => {
      pending = pending.length ? Buffer.concat([pending, chunk]) : chunk;
      let off = 0;
      while (pending.length - off >= frameBytes) {
        const frame = pending.subarray(off, off + frameBytes);
        if (prev) {
          let sum = 0;
          for (let i = 0; i < frameBytes; i++) sum += Math.abs(frame[i] - prev[i]);
          diffs.push(sum / frameBytes);
        }
        prev = new Uint8Array(frame); // 拷贝, 脱离 pending 底层
        off += frameBytes;
      }
      if (off) pending = Buffer.from(pending.subarray(off));
    });
    child.stderr.on("data", (d: Buffer) => { stderr += d.toString(); });
    child.on("error", reject);
    child.on("close", (code) => {
      if (code !== 0) reject(new Error(`ffmpeg failed (${code}): ${stderr}`));
      else resolve({ diffs, fps: meta.fps, width: meta.width, height: meta.height });
    });
  });
}

/** 完整 Step 1: 抽帧差 → 状态机切分。 */
export async function detectMotionSegments(
  videoPath: string,
  ffmpegPath = "ffmpeg",
  ffprobePath = "ffprobe",
): Promise<{ segments: Segment[]; diffs: number[]; fps: number; totalFrames: number }> {
  const { diffs, fps } = await computeFrameDiffs(videoPath, ffmpegPath, ffprobePath);
  const segments = segmentFromDiffs(diffs, fps);
  return { segments, diffs, fps, totalFrames: diffs.length + 1 };
}

/** 打印运动时间线 (对应 motion_detect.printTimeline)。 */
export function printTimeline(segments: Segment[]): void {
  let moveCount = 0;
  console.log("\n" + "=".repeat(70));
  console.log("Motion Timeline");
  console.log("=".repeat(70));
  for (const seg of segments) {
    const duration = (seg.endTime ?? 0) - (seg.startTime ?? 0);
    const frameCount = seg.endFrame - seg.startFrame;
    const label = seg.type === "MOVING" ? `  MOVE #${String(++moveCount).padStart(2)}` : "  STATIC   ";
    console.log(
      `${label}  [${(seg.startTime ?? 0).toFixed(2)}s - ${(seg.endTime ?? 0).toFixed(2)}s]` +
      `  (${duration.toFixed(2)}s, ${frameCount} frames)`,
    );
  }
  console.log("=".repeat(70));
  console.log(`Total moves detected: ${moveCount}`);
  console.log(`Total segments: ${segments.length}`);
}

// CLI: tsx src/motion-detect.ts <video.mp4>
if (process.argv[1] && process.argv[1] === fileURLToPath(import.meta.url)) {
  const videoPath = process.argv[2];
  if (!videoPath) {
    console.error("Usage: tsx src/motion-detect.ts <video.mp4>");
    process.exit(1);
  }
  console.log(`Analyzing: ${videoPath}`);
  detectMotionSegments(videoPath)
    .then(({ segments, diffs, fps, totalFrames }) => {
      console.log(`Computed ${diffs.length} frame diffs (${totalFrames} frames, ${fps.toFixed(2)} fps)`);
      printTimeline(segments);
    })
    .catch((e) => {
      console.error(e instanceof Error ? e.message : e);
      process.exit(1);
    });
}

File: reconer/motion_detect.py
"""
运动检测模块 — 从魔方视频中切分出 静止/转动 片段。

原理：
  对魔方 ROI 区域计算相邻帧的灰度绝对差的均值。
  当差值超过阈值时视为"运动帧"，低于阈值时视为"静止帧"。
  通过状态机将连续帧聚合为运动片段。
"""

import cv2
import numpy as np
import sys
import os


# ============================================================
# 配置参数
# ============================================================

# 魔方 ROI 区域 (归一化坐标, 相对于整帧)
# NOTE: 从截图观察, 魔方大约在画面中央偏上
ROI_X1, ROI_Y1 = 0.25, 0.15  # 左上角
ROI_X2, ROI_Y2 = 0.75, 0.85  # 右下角

# 运动检测阈值
MOTION_THRESHOLD = 5.0    # 高于此值 → 运动帧
STATIC_THRESHOLD = 3.0    # 低于此值 → 静止帧 (有回滞, 防抖动)

# 最短运动片段长度 (帧数), 过短的视为噪声
MIN_MOVE_FRAMES = 3

# 最短静止间隔 (帧数), 过短的静止视为同一个动作的一部分
MIN_STATIC_FRAMES = 5


def getRoi(frame, h, w):
    """从帧中截取 ROI 区域。"""
    x1 = int(w * ROI_X1)
    y1 = int(h * ROI_Y1)
    x2 = int(w * ROI_X2)
    y2 = int(h * ROI_Y2)
    return frame[y1:y2, x1:x2]


def computeFrameDiff(prevGray, currGray, h, w):
    """计算两帧 ROI 区域的灰度绝对差均值。"""
    prevRoi = getRoi(prevGray, h, w)
    currRoi = getRoi(currGray, h, w)
    diff = cv2.absdiff(prevRoi, currRoi)
    return np.mean(diff)


def detectMotionSegments(videoPath):
    """
    检测视频中的运动片段。

    返回:
        segments: list of dict, 每个元素:
            {
                'type': 'STATIC' | 'MOVING',
                'startFrame': int,
                'endFrame': int,
                'startTime': float,
                'endTime': float,
            }
        fps: float
        totalFrames: int
    """
    cap = cv2.VideoCapture(videoPath)
    if not cap.isOpened():
        print(f"Error: cannot open {videoPath}")
        sys.exit(1)

    fps = cap.get(cv2.CAP_PROP_FPS)
    totalFrames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    h = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    w = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))

    print(f"Video: {w}x{h}, {fps:.2f} fps, {totalFrames} frames, {totalFrames/fps:.2f}s")

    # 逐帧计算帧间差
    ret, prevFrame = cap.read()
    if not ret:
        print("Error: cannot read first frame")
        sys.exit(1)

    prevGray = cv2.cvtColor(prevFrame, cv2.COLOR_BGR2GRAY)
    # 轻度模糊去噪
    prevGray = cv2.GaussianBlur(prevGray, (5, 5), 0)

    diffs = []  # 每帧的差值

    frameIdx = 1
    while True:
        ret, frame = cap.read()
        if not ret:
            break

        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        gray = cv2.GaussianBlur(gray, (5, 5), 0)

        diff = computeFrameDiff(prevGray, gray, h, w)
        diffs.append(diff)

        prevGray = gray
        frameIdx += 1

    cap.release()
    print(f"Computed {len(diffs)} frame diffs")

    # 用状态机将帧分为 STATIC / MOVING
    states = []  # 每帧的状态标签
    state = "STATIC"

    for diff in diffs:
        if state == "STATIC":
            if diff > MOTION_THRESHOLD:
                state = "MOVING"
        else:  # MOVING
            if diff < STATIC_THRESHOLD:
                state = "STATIC"
        states.append(state)

    # 聚合为片段
    rawSegments = []
    if states:
        currState = states[0]
        startIdx = 0
        for i in range(1, len(states)):
            if states[i] != currState:
                rawSegments.append({
                    'type': currState,
                    'startFrame': startIdx + 1,  # +1 因为 diffs 从第1帧开始
                    'endFrame': i,
                })
                currState = states[i]
                startIdx = i
        # 最后一段
        rawSegments.append({
            'type': currState,
            'startFrame': startIdx + 1,
            'endFrame': len(states),
        })

    # 过滤：移除过短的运动片段 (视为噪声)
    filtered = []
    for seg in rawSegments:
        duration = seg['endFrame'] - seg['startFrame']
        if seg['type'] == 'MOVING' and duration < MIN_MOVE_FRAMES:
            # 把过短的运动片段变为静止
            seg['type'] = 'STATIC'
        filtered.append(seg)

    # 合并相邻的同类型片段
    merged = [filtered[0]]
    for seg in filtered[1:]:
        if seg['type'] == merged[-1]['type']:
            merged[-1]['endFrame'] = seg['endFrame']
        else:
            merged.append(seg)

    # 过滤：移除过短的静止间隔 (合并为一个连续运动)
    filtered2 = []
    for seg in merged:
        duration = seg['endFrame'] - seg['startFrame']
        if seg['type'] == 'STATIC' and duration < MIN_STATIC_FRAMES and filtered2:
            # 把过短的静止间隔变为运动
            seg['type'] = 'MOVING'
        filtered2.append(seg)

    # 再次合并
    final = [filtered2[0]]
    for seg in filtered2[1:]:
        if seg['type'] == final[-1]['type']:
            final[-1]['endFrame'] = seg['endFrame']
        else:
            final.append(seg)

    # 添加时间信息
    for seg in final:
        seg['startTime'] = seg['startFrame'] / fps
        seg['endTime'] = seg['endFrame'] / fps

    return final, diffs, fps, totalFrames


def printTimeline(segments, fps):
    """打印运动时间线。"""
    moveCount = 0
    print("\n" + "=" * 70)
    print("Motion Timeline")
    print("=" * 70)

    for seg in segments:
        duration = seg['endTime'] - seg['startTime']
        frameCount = seg['endFrame'] - seg['startFrame']

        if seg['type'] == 'MOVING':
            moveCount += 1
            label = f"  MOVE #{moveCount:2d}"
        else:
            label = "  STATIC   "

        print(f"{label}  [{seg['startTime']:6.2f}s - {seg['endTime']:6.2f}s]"
              f"  ({duration:.2f}s, {frameCount} frames)")

    print("=" * 70)
    print(f"Total moves detected: {moveCount}")
    print(f"Total segments: {len(segments)}")


def saveDiffPlot(diffs, fps, outputPath):
    """将帧间差值曲线保存为图片 (纯 OpenCV 实现, 无需 matplotlib)。"""
    plotW, plotH = 1200, 400
    img = np.ones((plotH, plotW, 3), dtype=np.uint8) * 255

    n = len(diffs)
    if n == 0:
        return

    maxDiff = max(diffs) * 1.1 if max(diffs) > 0 else 1.0

    # 绘制阈值线
    threshY = int(plotH - (MOTION_THRESHOLD / maxDiff) * (plotH - 40) - 20)
    cv2.line(img, (0, threshY), (plotW, threshY), (0, 0, 255), 1)
    cv2.putText(img, f"threshold={MOTION_THRESHOLD}", (plotW - 200, threshY - 5),
                cv2.FONT_HERSHEY_SIMPLEX, 0.4, (0, 0, 255), 1)

    # 绘制差值曲线
    points = []
    for i, d in enumerate(diffs):
        x = int(i / n * plotW)
        y = int(plotH - (d / maxDiff) * (plotH - 40) - 20)
        points.append((x, y))

    for i in range(1, len(points)):
        color = (0, 0, 200) if diffs[i] > MOTION_THRESHOLD else (200, 200, 200)
        cv2.line(img, points[i-1], points[i], color, 1)

    # X 轴时间标注
    for sec in range(0, int(n / fps) + 1, 2):
        x = int(sec * fps / n * plotW)
        cv2.putText(img, f"{sec}s", (x, plotH - 5),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.35, (100, 100, 100), 1)

    cv2.imwrite(outputPath, img)
    print(f"\nDiff plot saved to: {outputPath}")


def saveStableFrames(videoPath, segments, outputDir):
    """
    提取每个运动片段前后的静止帧。
    保存为 move_N_before.png / move_N_after.png。
    """
    os.makedirs(outputDir, exist_ok=True)
    cap = cv2.VideoCapture(videoPath)
    fps = cap.get(cv2.CAP_PROP_FPS)

    moveIdx = 0
    for i, seg in enumerate(segments):
        if seg['type'] != 'MOVING':
            continue
        moveIdx += 1

        # before: 取前一个静止片段的最后一帧
        if i > 0 and segments[i-1]['type'] == 'STATIC':
            beforeFrame = segments[i-1]['endFrame'] - 1
        else:
            beforeFrame = seg['startFrame']

        # after: 取后一个静止片段的第一帧 + 几帧偏移 (等手稳定)
        if i + 1 < len(segments) and segments[i+1]['type'] == 'STATIC':
            afterFrame = segments[i+1]['startFrame'] + 3
        else:
            afterFrame = seg['endFrame']

        for label, fnum in [('before', beforeFrame), ('after', afterFrame)]:
            cap.set(cv2.CAP_PROP_POS_FRAMES, fnum)
            ret, frame = cap.read()
            if ret:
                path = os.path.join(outputDir, f"move_{moveIdx:02d}_{label}.png")
                cv2.imwrite(path, frame)

    cap.release()
    print(f"\nStable frames saved to: {outputDir}/ ({moveIdx} moves)")


if __name__ == "__main__":
    videoPath = sys.argv[1] if len(sys.argv) > 1 else "test.mp4"

    print(f"Analyzing: {videoPath}")
    segments, diffs, fps, totalFrames = detectMotionSegments(videoPath)

    printTimeline(segments, fps)
    saveDiffPlot(diffs, fps, "diff_plot.png")
    saveStableFrames(videoPath, segments, "stable_frames")
